pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -0,0 +1,301 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/tenant"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/sync/errgroup"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/distributor"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storage/sharding"
+	"github.com/grafana/mimir/pkg/util"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+type shardActiveSeriesMiddleware struct {
+	upstream http.RoundTripper
+	limits   Limits
+	logger   log.Logger
+}
+
+func newShardActiveSeriesMiddleware(upstream http.RoundTripper, limits Limits, logger log.Logger) http.RoundTripper {
+	return &shardActiveSeriesMiddleware{
+		upstream: upstream,
+		limits:   limits,
+		logger:   logger,
+	}
+}
+
+func (s *shardActiveSeriesMiddleware) RoundTrip(r *http.Request) (*http.Response, error) {
+	const defaultNumShards = 1
+
+	spanLog, ctx := spanlogger.NewWithLogger(r.Context(), s.logger, "shardActiveSeries.RoundTrip")
+	defer spanLog.Finish()
+
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	numShards := setShardCountFromHeader(defaultNumShards, r, spanLog)
+
+	if numShards < 2 {
+		spanLog.DebugLog("msg", "query sharding disabled for request")
+		return s.upstream.RoundTrip(r)
+	}
+
+	if maxShards := s.limits.QueryShardingMaxShardedQueries(tenantID); numShards > maxShards {
+		return nil, apierror.New(
+			apierror.TypeBadData,
+			fmt.Sprintf("shard count %d exceeds allowed maximum (%d)", numShards, maxShards),
+		)
+	}
+
+	selector, err := parseSelector(r)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	spanLog.DebugLog(
+		"msg", "sharding active series query",
+		"shardCount", numShards, "selector", selector.String(),
+	)
+
+	reqs, err := buildShardedRequests(ctx, r, numShards, selector)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeInternal, err.Error())
+	}
+
+	resp, err := doShardedRequests(ctx, reqs, s.upstream)
+	if err != nil {
+		if errors.Is(err, distributor.ErrResponseTooLarge) {
```

Comment:
How is this error instantiatead in query-frontend's middleware?

Reply:
This was actually broken, I need to figure out a way to test this.